### PR TITLE
Refactors Suiciding Variable Into Trait

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

Firstly, this var was on `/mob`, even though only `/mob/living` and `/mob/dead` could have ever used it, so who knows how much needless memory it was consuming on stuff such as `oranges_ear` that would never ever ever use something like this.

Edit: okay instead of memory it just polluted variable edit windows for all /mob when it didn't need to. I like having a slim VV window

Secondly, it's a technical improvement over the previous system as we are able to "track" where a suicide originates from, and how we can track that from mob-to-mob-to-mob. Previously, the boolean `suiciding` would only inform us if they had ever been connected to a mob that had ever committed suicide, but now we are able to precisely determine which mob gave them the trait that they must now apparently bear until the round restarts.

## Why It's Good For The Game

Less memory usage, more indepth ability to track suicides in case you really need that dexterity. Currently no implemented code could benefit from using it, but it would be pretty neat if someone could figure out a way to have someone be guilt-tripped whenever they look into a mirror and seeing the reflection of their past life? This PR won't actually help you code that and it'll probably require a bit more work, but it's a possibility of some cool interactions you can do when you have this information available to you.

image

## Changelog

:cl:
refactor: Some aspects of how we track suicides from your living mob to your observer have changed- please do let us know if anything has broken via a GitHub Issue Report.
/:cl:

There's probably some technical improvements that can be made in some parts of the code I reworked to accommodate this change, do let me know if you spot any easy ones (or fuckups). a lot of excess comes from the fact that any step in the TRAIT framework trusts that you are passing in a valid datum (or subtype) so that's a thing

### CODE REVIEW

File: code/modules/client/verbs/suicide.dm [hunk 1]
```diff
@@ -16,13 +16,15 @@
...
 	final_checkout()
 
-/// Proc that handles changing the suiciding var on the mob in question, as well as additional operations to ensure that everything goes smoothly when we're certain that this person is going to kill themself.
-/// suicide_state is a boolean, to match the suiciding/suicided var.
-/mob/proc/set_suicide(suicide_state)
-	suiciding = suicide_state
+/// Proc that handles adding the TRAIT_SUICIDED on the mob in question, as well as additional operations to ensure that everything goes smoothly when we're certain that this person is going to kill themself.
+/// suicide_state is a boolean, and we handle adding/removing the trait in question. Have the trait function reference this mob as the source if we want to do in-depth tracking of where a suicided trait comes from.
+/// For example, the /mob/dead/observer that will inevitably come from the suicidee will inherit the suicided trait upon creation, and keep this reference. Handy for doing checking should we need it.
+/mob/living/proc/set_suicide(suicide_state)
```
Comment (san7890) [inline code/modules/client/verbs/suicide.dm:22]: this was only ever invoked via handle_suicide() iirc

File: code/modules/mob/mob_defines.dm [hunk 2]
```diff
@@ -65,9 +65,6 @@
 	///Cursor icon used when holding shift over things
 	var/examine_cursor_icon = 'icons/effects/mouse_pointers/examine_pointer.dmi'
 
-	///Whether this mob has or is in the middle of committing suicide.
-	var/suiciding = FALSE
-
 	/// Whether a mob is alive or dead. TODO: Move this to living - Nodrak (2019, still here)
```
Comment (san7890) [inline code/modules/mob/mob_defines.dm:68]: half tempted to write (2023, still here) but

**Comment by Mothblocks** [comment]
Only because you asked and it's fun knowledge, this PR almost certainly takes *more* memory than the `suiciding`
 var. A variable in its default state is completely free.

File: code/modules/mob/mob_defines.dm (hunk 2 above)
Comment (Mothblocks) [reply code/modules/mob/mob_defines.dm:68]: I'd rather you replace it with "Ideally this is on living instead, but an enormous amount of code would need to change"

Comment (san7890) [reply code/modules/mob/mob_defines.dm:68]: I'll chalk it up to "out of scope for the purposes of this PR", just think it's funny that this var was originally put in such close proximity to this

**Comment by san7890** [comment]
> Only because you asked and it's fun knowledge, this PR almost certainly takes *more* memory than the `suiciding`
> 
>  var. A variable in its default state is completely free. It's a good PR regardless.

Good thing I wasn't hedging my bets on that, and also a good thing to know.

**Comment by ZephyrTFA** [comment]
traits don't really have that much memory overhead and this being a trait is better than an off/on flag

**Pull request merged**